src/container/srv_target.c

```diff
@@ -2324,58 +2347,150 @@ ds_cont_ec_eph_insert(struct ds_pool *pool, uuid_t cont_uuid, int tgt_idx,
-	new_eph->ce_ephs[tgt_idx] = 0;
-	*epoch_p = &new_eph->ce_ephs[tgt_idx];
+	new_eph->ce_ephs[tgt_idx].ec_last_aggregation_epoch = d_hlc_get();
+	new_eph->ce_ephs[tgt_idx].dtx_commit_min_epoch = d_hlc_get();
+	*ec_epoch_p = &new_eph->ce_ephs[tgt_idx].ec_last_aggregation_epoch;
+	*dtx_commit_epoch_p = &new_eph->ce_ephs[tgt_idx].dtx_commit_min_epoch;
 out:
 	return rc;
 }
 
 int
-ds_cont_ec_eph_delete(struct ds_pool *pool, uuid_t cont_uuid, int tgt_idx)
+ds_cont_track_eph_delete(struct ds_pool *pool, uuid_t cont_uuid, int tgt_idx)
 {
-	struct cont_ec_eph	*ec_eph;
+	struct cont_track_eph	*delete_eph;
 
 	D_ASSERT(dss_get_module_info()->dmi_xs_id == 0);
-	ec_eph = cont_ec_eph_lookup(&pool->sp_ec_ephs_list, cont_uuid);
-	if (ec_eph == NULL)
+	delete_eph = cont_track_eph_lookup(&pool->sp_track_ephs_list, cont_uuid);
+	if (delete_eph == NULL)
 		return 0;
 
-	D_ASSERT(tgt_idx < ec_eph->ce_ephs_cnt);
-	D_ASSERT(ec_eph->ce_ref > 0);
-	ec_eph->ce_ref--;
+	D_ASSERT(tgt_idx < delete_eph->ce_ephs_cnt);
+	D_ASSERT(delete_eph->ce_ref > 0);
+	delete_eph->ce_ref--;
 	D_DEBUG(DB_MD, DF_UUID "delete %d tgt ref %d.\n",
-		DP_UUID(cont_uuid), tgt_idx, ec_eph->ce_ref);
+		DP_UUID(cont_uuid), tgt_idx, delete_eph->ce_ref);
+	return 0;
+}
+
+static void
+cont_track_eph_destroy(struct cont_track_eph *delete_eph)
+{
+	d_list_del(&delete_eph->ce_list);
+	D_FREE(delete_eph->ce_ephs);
+	D_FREE(delete_eph);
+}
+
+struct track_eph_ult_arg {
+	struct ds_pool	*pool;
+	uuid_t		cont_uuid;
+	uint32_t	tgt_idx;
+	daos_epoch_t	*ec_eph;
+	daos_epoch_t	*dtx_eph;
+};
+
+static	int
+cont_track_eph_fini_ult(void *data)
+{
+	struct track_eph_ult_arg	*arg = data;
+	int				rc;
+
+	rc = ds_cont_track_eph_delete(arg->pool, arg->cont_uuid, arg->tgt_idx);
+	D_ASSERT(rc == 0);
 	return 0;
 }
 
 static void
-cont_ec_eph_destroy(struct cont_ec_eph *ec_eph)
+cont_tgt_track_eph_fini(struct ds_cont_child *cont_child)
 {
-	d_list_del(&ec_eph->ce_list);
-	D_FREE(ec_eph->ce_ephs);
-	D_FREE(ec_eph);
+	struct track_eph_ult_arg	arg;
+
+	if (cont_child->sc_ec_query_agg_eph == NULL)
+		return;
+
+	arg.pool = cont_child->sc_pool->spc_pool;
+	uuid_copy(arg.cont_uuid, cont_child->sc_uuid);
+	arg.tgt_idx = dss_get_module_info()->dmi_tgt_id;
+	dss_ult_execute(cont_track_eph_fini_ult, &arg, NULL, NULL, DSS_XS_SYS, 0, 0);
+
+	cont_child->sc_ec_query_agg_eph = NULL;
+	cont_child->sc_dtx_commit_eph = NULL;
+}
+
+static int
+cont_track_eph_init_ult(void *data)
+{
+	struct track_eph_ult_arg *arg = data;
+	int rc;
+
+	rc = ds_cont_track_eph_insert(arg->pool, arg->cont_uuid, arg->tgt_idx,
+				      &arg->ec_eph, &arg->dtx_eph);
+	return rc;
+}
+
+static void
+cont_tgt_track_eph_init_ult(void *data)
+{
+	struct ds_cont_child		*cont_child = data;
+	struct track_eph_ult_arg	arg;
+	int				rc;
+
+	arg.pool = cont_child->sc_pool->spc_pool;
+	uuid_copy(arg.cont_uuid, cont_child->sc_uuid);
+	arg.tgt_idx = dss_get_module_info()->dmi_tgt_id;
+	rc = dss_ult_execute(cont_track_eph_init_ult, &arg, NULL, NULL, DSS_XS_SYS,
```

Comment:
just confirm recreate ULT to run on SYS XS is just to avoid taking lock? or seems ds_cont_track_eph_insert() can be executed on self XS?

Reply:
init_ult will check and create the ec array for all other streams,  so better to do it on sys XS.